lib/IRGen/DebugTypeInfo.h

```diff
@@ -47,51 +48,57 @@ class DebugTypeInfo {
-  DebugTypeInfo(swift::Type Ty, const TypeInfo &Info, DeclContext *DC);
-  DebugTypeInfo(TypeDecl *Decl, const TypeInfo &Info);
-  DebugTypeInfo(ValueDecl *Decl, llvm::Type *StorageType, Size size,
-                Alignment align);
-  DebugTypeInfo(VarDecl *Decl, swift::Type Ty, const TypeInfo &Info,
-                bool Unwrap);
-  DebugTypeInfo(VarDecl *Decl, swift::Type Ty,
-                llvm::Type *StorageType, Size size,
-                Alignment align);
+  DebugTypeInfo(DeclContext *DC, swift::Type Ty, llvm::Type *StorageTy,
+                Size SizeInBytes, Alignment AlignInBytes);
+  /// Create type for a local variable.
+  static DebugTypeInfo getLocalVariable(DeclContext *DeclCtx, VarDecl *Decl,
+                                        swift::Type Ty, const TypeInfo &Info,
+                                        bool Unwrap);
+  /// Create type for an artificial metadata variable.
+  static DebugTypeInfo getMetadata(swift::Type Ty, llvm::Type *StorageTy,
+                                   Size size, Alignment align);
+  /// Create a standalone type from a TypeInfo object.
+  static DebugTypeInfo getFromTypeInfo(DeclContext *DC, swift::Type Ty,
+                                       const TypeInfo &Info);
+  /// Global variables.
+  static DebugTypeInfo getGlobal(SILGlobalVariable *GV, llvm::Type *StorageType,
+                                 Size size, Alignment align);
+  /// ObjC classes.
+  static DebugTypeInfo getObjCClass(ClassDecl *theClass,
+                                    llvm::Type *StorageType, Size size,
+                                    Alignment align);
+
   TypeBase *getType() const { return Type; }
 
   TypeDecl *getDecl() const;
   DeclContext *getDeclContext() const { return DeclCtx; }
 
   void unwrapLValueOrInOutType() {
     Type = Type->getLValueOrInOutObjectType().getPointer();
-  }
-
-  // Determine whether this type is an Archetype itself.
-  bool isArchetype() const {
-    return Type->getLValueOrInOutObjectType()->getDesugaredType()->getKind() ==
-           TypeKind::Archetype;
-  }
-
-  /// LValues, inout args, and Archetypes are implicitly indirect by
-  /// virtue of their DWARF type.
-  bool isImplicitlyIndirect() const {
-    return Type->isLValueType() || isArchetype() ||
-           (Type->getKind() == TypeKind::InOut);
-  }
-
-  bool isNull() const { return Type == nullptr; }
-  bool operator==(DebugTypeInfo T) const;
-  bool operator!=(DebugTypeInfo T) const;
-
-  void dump() const;
+    }
+
+    // Determine whether this type is an Archetype itself.
+    bool isArchetype() const {
+      return Type->getLValueOrInOutObjectType()
+                 ->getDesugaredType()
+                 ->getKind() == TypeKind::Archetype;
+    }
+
+    /// LValues, inout args, and Archetypes are implicitly indirect by
+    /// virtue of their DWARF type.
+    bool isImplicitlyIndirect() const {
+      return Type->isLValueType() || isArchetype() ||
+             (Type->getKind() == TypeKind::InOut);
```

Comment:
Side note: I would like to follow-up on this separately from this patch, this line only had clang-format changes on it and was otherwise unchanged.
That said, I'm very interested in cleaning all of this up! Thanks for looking.